### ISSUE TITLE
SecureRandom is slow

### ISSUE THREAD

**Issue by 98-f355-f1** [oracle/truffleruby#1395]
```ruby
require benchmark
require securerandom

N = 100_000
RANGE = 0...999_999

Benchmark.bm do |x|
  x.report('rand()') { N.times { rand(RANGE) } }
  x.report('SecureRandom.rand') { N.times { SecureRandom.uuid } }
  x.report('SecureRandom.random_bytes') { N.times { SecureRandom.random_bytes(16) } }
end
```

I finally got all of the major Ruby versions going; So I ran this first test:

MRI RUBY

```
       user     system      total        real
rand()  0.031250   0.000000   0.031250 (  0.017138)
SecureRandom.rand  0.359375   0.156250   0.515625 (  0.517790)
SecureRandom.random_bytes  0.062500   0.125000   0.187500 (  0.193459)
```

JRUBY

```
       user     system      total        real
rand()  0.470000   0.050000   0.520000 (  0.154377)
SecureRandom.rand  0.560000   0.030000   0.590000 (  0.202068)
SecureRandom.random_bytes  0.410000   0.000000   0.410000 (  0.138920)
```

CRYSTAL

```
                                  user     system      total        real
rand()                        0.000000   0.000000   0.000000 (  0.000685)
Random::Secure.rand           0.010000   0.160000   0.170000 (  0.163817)
Random::Secure.random_bytes   0.030000   0.120000   0.150000 (  0.161642)
```

and then TRUFFLE

```
       user     system      total        real
rand()  0.718750   0.500000   1.218750 (  0.505922)
SecureRandom.rand 14.093750  57.312500  71.406250 ( 66.696175)
SecureRandom.random_bytes  9.234375  58.765625  68.000000 ( 66.3593)
```

Looking at these numbers: any possible explanations as to why the TRUFFLE is so slow with this?

I hopefully followed all the instructions for install TRUFFLE.  I read there is a flag to optimize the compiler?

Any suggestions?

Thanks

**Comment by nirvdrum** [comment]
Thanks for filing this performance issue! Off-hand, I suspect there's a problem with the way we're running the native OpenSSL extension. Perhaps complicating the issue is our interpreter performance with native extensions is known to be slow. The benchmark code as you've written it doesn't have a warm-up phase.

I realize that's not a very satisfactory answer though. We should be faster in the interpreter as well. This will also get better when we update the standard library to 2.5, since we'll end up reading from `/dev/urandom` in that case.

If you get a chance, could you please edit the issue to include the version numbers of each of the Ruby implementations you've tested?

**Comment by 98-f355-f1** [comment]
I have the latest releases of all the Rubies:
MRI = 2.5.1
JRuby = 9.2.1
Crystal = .25
Truffle = 1.0 on rc4 Graal

Yes, this was just an impromptu test, mainly to check that all of the Rubies were performing ok. I didn't have the Benchmark-ips version of the Benchmark installed for all Rubies at that time. I do now, I can give an update with those numbers.

So, far Truffle has done extremely well in other tests. This test was the outlier for Truffle performance, a clear indication that possibly as you noted that the module isn't setup optimally yet.

Thanks

**Comment by nirvdrum** [comment]
Thanks for the update.

So, there will be a bit of a difference because of the change in which random number generators are used first that occurred in Ruby 2.5. TruffleRuby still targets Ruby 2.4. But, I think we can generally do better anyway. Those numbers you presented are quite bad.

**Comment by nirvdrum** [comment]
This has been fixed in 5160de7b1d7d6dc1be5a7e81850e8ee13fe066c1. Unfortunately, I don't think it'll be available in our next release (1.0.0 RC 5) just based on the release schedule. But it should be available in the release after that (tentatively 1.0.0 RC 6).

On my machine, with TruffleRuby 1.0.0 RC 3 I had the following results:

```
> ruby -v securerandom-benchmark.rb 
truffleruby 1.0.0-rc3, like ruby 2.4.4, GraalVM CE Native [x86_64-linux]
       user     system      total        real
rand()  0.530894   0.053839   0.584733 (  0.264590)
SecureRandom.rand  9.015708   0.972648   9.988356 (  6.923810)
SecureRandom.random_bytes  5.286068   0.820912   6.106980 (  4.598315)
```

Now I'm seeing:

```
> ./bin/native-ruby -v ~/securerandom-benchmark.rb 
truffleruby 1.0.0-rc5-dev-7372fe92, like ruby 2.4.4, GraalVM CE Native [x86_64-linux]
       user     system      total        real
rand()  0.524400   0.044191   0.568591 (  0.243052)
SecureRandom.rand  0.763358   0.103915   0.867273 (  0.490850)
SecureRandom.random_bytes  0.093558   0.084167   0.177725 (  0.158959)
```

There's still some opportunity for improvement, particularly with `rand`. I didn't really look at that one. But `SecureRandom` is looking much better now.

I'm going to close the issue out, but please re-open if you think I've closed it prematurely.

**Comment by nirvdrum** [comment]
@98-f355-f1TruffleRuby 1.0.0 RC6 shipped with the fix I mentioned for `SecureRandom`. Can you please try your benchmark script on your machine and verify whether the problem has been fixed? Thanks!
